Migração de NFe para .net Standard (.netCore + .netFramework)

danilobreda opened the pull request:
#1001 

Detalhe importante: alguns arquivos foram convertidos de unicode para UTF8, por isso o numero de changed files é grande, além claro de movimentações de classes para shared projects...

danilobreda committed:
criado projetos

danilobreda committed:
movido servicos do nfe

danilobreda committed:
movido projeto wsdl, NFe.Utils.Annotations arrumado, colocado em Shared.Nfe.Utils

danilobreda committed:
criado shared project do danfe

danilobreda committed:
alteracoes de danfe iniciais

danilobreda committed:
conversão para utf8
Criado o Shared para Danfe, e nova classe DanfeHelper para centralizar códigos duplicados

danilobreda committed:
Migrado DANFE (NFe e NFe Eventos) em .net Standard, falta NFCe.
Geração de PDF, HTML ou IMAGEM PNG em arrays de bytes.
Feito console com a implementação das novas classes.

danilobreda committed:
arrumado erro de UTF8

danilobreda committed:
algumas modificacoes no danfe para carregamento de confiuracoes e inicio do dev de console do NFE

danilobreda committed:
pasta conversão

danilobreda committed:
adicionado retorno em stream do danfe

danilobreda committed:
Retirado algumas classes do projeto shared onde não funcionam no .netstandard, para melhor entendimento e separação da etapa de conversão.
As classes foram transferidas do WSDL SHARED para classe especifica do .net framework

danilobreda committed:
feito a primeiro exemplo de conversao do wsdl

danilobreda committed:
comentado elementos do ServicoNfeFactory para testes e migração

danilobreda committed:
Arquivos convertidos para UTF8 BOM

danilobreda committed:
feito a parte de STATUS de NFE, com alguns estados não implementado

danilobreda committed:
parte inicial autorizacao

danilobreda committed:
terminado a parte de autorizacao

danilobreda committed:
feito nfceCsc

danilobreda committed:
implementação/migracao do consulta protocolo

danilobreda committed:
criado classes de distribuicaodfe e download (ainda não implementadas)

danilobreda committed:
Migrado eventos da nfe

danilobreda committed:
implementado inutilizacao (varias versoes) e recepcao (v2.0)

danilobreda committed:
implementado e migrado RetAutorizacao

danilobreda committed:
arrumado alguns problemas de identação

danilobreda commented:
Da mandeira que foi feita o novo relatorio de FastReport para o Danfe, é necessário que tenha 2 arquivos frxs em cada projeto (um no .net framework antigo e outro no .net core) Isso duplica o relatório e dificulta mudanças ao longo do tempo, Vou tentar jogar os relatorios no projeto Shared e fazer alteracoes em runtime do frx na hora da impressão no .net core!

marcosgerene commented:
@danilobreda  
> Da mandeira que foi feita o novo relatorio de FastReport para o Danfe, é necessário que tenha 2 arquivos frxs em cada projeto (um no .net framework antigo e outro no .net core) Isso duplica o relatório e dificulta mudanças ao longo do tempo, Vou tentar jogar os relatorios no projeto Shared e fazer alteracoes em runtime do frx na hora da impressão no .net core!

Eu tenho clientes cujo os relatórios são modificados sob demanda. Como ficaria essa alteração em runtime nesse caso? Acho isso um pouco perigoso...

danilobreda committed:
Retirado os relatorios dos projetos especificos e colocados em projeto shared.
Foi criado um metodo que adiciona e altera assemblies no caso apenas do dotnet core (ArrumaReferencedAssembliesCore).

danilobreda commented:
@marcosgerene Fiz as modificações para você dar uma olhada... Eu achei melhor assim, para que fique apenas 1 relatório tanto nos projetos fast antigo e novo. O que o código do novo método ArrumaReferencedAssembliesCore faz, é adicionar referencias a um array do relatorio de .net standard. Apenas isso.
Não existe outra maneira, ou duplicamos relatório (o que eu acho pior) ou o código é um pouco inteligente para que possa rodar o relatório nos 2 (como o problema são assemblies achei mais tranquilo).
Porem no seu caso acredito que uma das plataformas deve ser escolhida. O net standard viabiliza tanto .net framework quanto o .net core e outros.
Talvez a ideia de deixar o Danfe em .net framework seja temporário... o futuro é .net standard pois já resolve o problema do .net framework!

danilobreda commented:
Vale lembrar que o relatório NFCe não está migrado para o novo danfe, pois não consigo testar por enquanto. O mesmo acontece no NFe, foi focado NFe, NFCe ainda não suporta .net standard!

marcosgerene commented:
@danilobreda 

Manter os dois relatórios não me parece tão ruim a curto prazo, as alterações que fazemos na Danfe são MUITO poucas, acho melhor a separação neste caso específico, caso contrário vai causar quebra para mais pessoas que já personalizaram de alguma forma.

danilobreda committed:
Revert "Retirado os relatorios dos projetos especificos e colocados em projeto shared."

This reverts commit d125d10902f5e9655593c16fca7be93f7162ab6a.

danilobreda committed:
alteracoes para resolver conflitos com alteracoes feitas no retrato a 2 dias.

danilobreda commented:
Verificar o conflito do NFeRetato, atualizei o arquivo para a ultima versão de 2 dias atras, sem sucesso :/

marcosgerene committed:
Merge branch 'master' into nfecoreport

marcosgerene commented:
Ajustei direto no github para tirar o conflito, nem sequer baixei no VS. Teste novamente o frx, por favor

danilobreda commented:
@marcosgerene  teste de impressão efetuados! tudo ok
Quanto ao código, alguma revisão de pontos especificos?

marcosgerene commented:
@danilobreda tive uma semana meio corrida e essa tende a ser tão quanto, assim que eu tiver tempo testo, mas acho que o @robertorp e o @adeniltonbs são mais indicados para esses testes, minha experiencia com .net standard e .net core são bem pequenas

danilobreda commented:
@marcosgerene  tem como gerar um nuget pre-release ou algo assim, para que possamos testar? Estou com medo de que aconteça algum problema a carregar o frx, pois testei local e deu problema.
Porem gerei um nuget de uma maneira bem errada..

danilobreda commented:
É parece quem temos um problema... o resource quando em nuget package, não consegue ser carregado... [/microsoft/msbuild/issues/2221](https://github.com/microsoft/msbuild/issues/2221)
### edit:
**O problema acontece apenas quando geramos um nuget do projeto zeus e usamos esse nuget.**

O resource do projeto nfe.danfe.fast.standard não consegue ser passado através de um nuget, via bug citado acima... A versão 3.0 está para resolver esse problema.

A saída é a injeção manual do usuário (ler o frx manualmente e transformar em bytes e injetar no projeto). Ou seja se eu tenho um _erp_ em .net core, foi ter que copiar os arquivos do .frx manualmente(do projeto do zeus) e colar no meu projeto, gerar esses arquivos junto com o publish via pasta comum, File.Open.. ler os bytes e passar como parametro no construtor da classe de carregamento do Danfe (Hoje tambem existe um overload, passando o caminho e ele ja carregar o .frx).

danilobreda committed:
removido arquivos que estavam em pasta errada do nfe.danfe.fast e mudado a maneira com que os arquivos de report são carregados via nuget como conteudo no projeto do danfe

marcosgerene commented:
@danilobreda 
>@marcosgerene tem como gerar um nuget pre-release ou algo assim, para que possamos testar? Estou com medo de que aconteça algum problema a carregar o frx, pois testei local e deu problema.
Porem gerei um nuget de uma maneira bem errada..

Quem controla o nuget é o @robertorp, não sei se tem como gerar um nuget desse branch especifico, deve ter... é util inclusive para testes

robertorp commented:
@danilobreda 
![image](https://user-images.githubusercontent.com/10251705/62320927-dfc75000-b477-11e9-978c-346fa7c77567.png)
![image](https://user-images.githubusercontent.com/10251705/62320939-e5249a80-b477-11e9-89bd-154ebba4b6fc.png)

como fazer funcionar? ai esta do mesmo jeito?

danilobreda committed:
alteracoes no DanfeFrNfe para que funcione corretamente e de erros mais amigaveis. O dev tambem pode criar uma classe abstrata  herdando de DanfeBase e gerar seu proprio código

danilobreda commented:
@robertorp ele não builda corretamente? não sei se é isso vi que o problema na imagem só acontece na aplicação que tem .net core, pois os outros projects são .net standard... acho que você tem que instalar/atualizar o sdk 2.2.x
Só baixar o SDK: [dotnet.microsoft.com/download/dotnet-core/2.2](https://dotnet.microsoft.com/download/dotnet-core/2.2)
(não sei se precisa baixar o runtime tambem, mas pelo que me lembro ele instala junto quando instala o sdk)
Temos que ver também se o método de geração dos nugets é diferente no .net standard...

danilobreda committed:
Arrumado arquivos frx ficava gerando sozinho no output, causando problemas para o dev ao usar a dll (e possivelmente o nuget), agora não gera mais.

danilobreda commented:
@robertorp  @marcosgerene Ocorreu um problema a nivel de certificado ao bater no Sefaz que acontece apenas em linux (no servidor por ex).
É importante utilizar a flag da configuracao **ValidarCertificadoDoServidor = false** para linux!
Hoje estou colocando em produção um mini sistema de gestão que utiliza as bibliotecas desse pullrequest.
Acredito que será um teste final para a aprovação do pullrequest!

marcosgerene commented:
@danilobreda 
>@robertorp @marcosgerene Ocorreu um problema a nivel de certificado ao bater no Sefaz que acontece apenas em linux (no servidor por ex).
É importante utilizar a flag da configuracao ValidarCertificadoDoServidor = false para linux!

Não ficou claro pra mim, "peguei o bonde andando", setando para false funciona no linux normalmente?

>Hoje estou colocando em produção um mini sistema de gestão que utiliza as bibliotecas desse pullrequest.
Acredito que será um teste final para a aprovação do pullrequest!

Excelente, um teste em produção controlado é sempre bem vindo

danilobreda commented:
@robertorp  O erro ocorreu ao tentar efetuar uma conexão com o Sefaz em produção (que é linux):

> Sem comunicação com o serviço NfeStatusServico:↵The SSL connection could not be established, see inner exception. The remote certificate is invalid according to the validation procedure.

Verifiquei que no windows funciona tranquilamente, apenas no linux dava o problema.
Verifiquei também uma issue antiga #822 especificamente [o seu comentario ](https://github.com/ZeusAutomacao/DFe.NET/issues/822#issuecomment-411741475)

[Nessa Issue](https://github.com/dotnet/corefx/issues/32751#issuecomment-497622676) Identifiquei que o linux trata diferente do windows a validação de um certificado. A solução é ignorar a invalidação desse certificado (retornando sempre true). Identifiquei que isso é já  é feito [nessa linha](https://github.com/ZeusAutomacao/DFe.NET/blob/4b7679eb8d65f9b3f8de6312e6685db1a14bf847/NFe.Servicos/ServicosNFe.cs#L108) no construtor da classe ServicosNFE.

Confirmei que setando false a variável ValidarCertificadoDoServidor resolve o problema no linux (docker ubuntu)!

marcosgerene commented:
@danilobreda 
>Verifiquei que no windows funciona tranquilamente, apenas no linux dava o problema.
Verifiquei também uma issue antiga #822 especificamente o seu comentario
>
>Nessa Issue Identifiquei que o linux trata diferente do windows a validação de um certificado. A solução é ignorar a invalidação desse certificado (retornando sempre true). Identifiquei que isso é já é feito nessa linha no construtor da classe ServicosNFE.
>
>Confirmei que setando false a variável ValidarCertificadoDoServidor resolve o problema no linux (docker ubuntu)!


Interessante, pode até ser que com a migração atual eu possa remover meu console...

aldoir commented:
> Confirmei que setando false a variável ValidarCertificadoDoServidor resolve o problema no linux (docker ubuntu)!

Isso está relacionado com a cadeia da Autoridade Certificadora do Brasil (ICP-BR) não existir no mundo Linux. Eles estão há 16 anos tentando a inclusão mas nunca atendem os requisitos estabelecidos para serem confiáveis. (Novela em https://bugzilla.mozilla.org/show_bug.cgi?id=438825).

Estou utilizando a parte de impressão de DANFE e DACTE com sucesso no Docker +  .NET Core 3.0 + Fast Reports Open Source. Foram necessários alguns ajustes no FRX pois aparentemente o Fast Reports Open Source (ou no Linux) tem algumas limitações, como só entender tamanho de fonte inteiro, mas no geral o resultado é excelente.

marcosgerene commented:
@aldoir
>Isso está relacionado com a cadeia da Autoridade Certificadora do Brasil (ICP-BR) não existir no mundo Linux. Eles estão há 16 anos tentando a inclusão mas nunca atendem os requisitos estabelecidos para serem confiáveis. (Novela em https://bugzilla.mozilla.org/show_bug.cgi?id=438825).
>
>Estou utilizando a parte de impressão de DANFE e DACTE com sucesso no Docker + .NET Core 3.0 + Fast Reports Open Source. Foram necessários alguns ajustes no FRX pois aparentemente o Fast Reports Open Source (ou no Linux) tem algumas limitações, como só entender tamanho de fonte inteiro, mas no geral o resultado é excelente.

Por favor, poste as alterações do frx para que possamos analisar

danilobreda commented:
@marcosgerene _Tive problemas também na geração do Danfe em linux, algumas fontes estouraram... o que vou fazer é diminuir a fonte para resolver igual a dica do @aldoir ._
**EDIT: Problema resolvido em comentarios abaixo...**

Outro problema foi a necessidade da DLL do gdiplus.dll através de uma exceção.
Adicionei o seguinte código no docker e funcionou!

> RUN apt-get update
> RUN apt-get install -y libgdiplus
> RUN ln -s /usr/lib/libgdiplus.so /usr/lib/gdiplus.dll

aldoir commented:
@danilobreda, para funcionar correntemente você precisa da fonte instalada no container também. No meu caso eu instalei a Times New Roman e padronizei todos os campos para usar ela (além de não usar casas decimais no tamanho da fonte). Caso a fonte não exista no sistema o Fast coloca a primeira disponível, que no caso do gdiplus é uma chamada DejaVu Sans e estoura tudo

danilobreda commented:
@aldoir  entendi, realmente vi que em uma maquina ubuntu local o pdf gero perfeito... acredito que nele esta instalado a fonte... voce tem o comando que é colocado no dockerfile para instalar essas fontes?
Fico grato pela ajuda 👍

aldoir commented:
> @aldoir entendi, realmente vi que em uma maquina ubuntu local o pdf gero perfeito... acredito que nele esta instalado a fonte... voce tem o comando que é colocado no dockerfile para instalar essas fontes?

Segue meu Dockerfile

```
FROM mcr.microsoft.com/dotnet/core/aspnet:3.0

RUN apt-get update \
    && apt-get install -y --no-install-recommends libgdiplus libc6-dev \
    && apt-get clean \
    && rm -rf /var/lib/apt/lists/*
RUN mkdir -p /usr/share/fonts/truetype/times
COPY fontes/* /usr/share/fonts/truetype/times/
```

Essa pasta **fontes** tem uma cópia de todos os arquivos .ttf da Times New Roman na pasta Windows, mas como são fontes com royalties não há um repositório online com as mesmas

```
ls -l fontes/*

fontes/timesbd.ttf
fontes/timesbi.ttf
fontes/timesi.ttf
fontes/times.ttf
```

Se essa vertente de .net core + linux engrenar talvez possamos fazer uma versão do FRX com fonte Open Source, como as do projeto fonts.google.com. Tem umas fontes com serifa (como a PT Serif) muito legíveis.

Outra coisa que eu fiz no meu projeto foi exportar para HTML e converter para PDF em outro microserviço (https://github.com/thecodingmachine/gotenberg), pois a exportação para PDF do Fast Reports OpenSource gera um JPEG estático, a DANFE fica com 350Kb sem possibilidade de CTRL+C / CTRL + V

robertorp commented:
Vou analisar o pull esse final de semana, se possível corrigir os conflitos que estão dando.

danilobreda commented:
@robertorp talvez teríamos que fazer um readme ou algum texto para os usuários que forem utilizar para linux, como viu existem alguns problemas como a necessidade da biblioteca libgdiplus no linux e as fontes de texto. Isso pode ser feito também posteriormente...

marcosgerene commented:
@danilobreda @aldoir 

Vocês já estão fazendo os testes em produção? poderiam demonstrar de alguma forma os resultados?

Se mostrassem rodando em ambiente Linux e Windows lado a lado seria lindo rs

aldoir commented:
> @danilobreda @aldoir
> 
> Se mostrassem rodando em ambiente Linux e Windows lado a lado seria lindo rs

Eu estou usando com .NET Core 3.0, mas meu uso é bem restrito, pois quando iniciei meu projeto ainda não havia esse movimento de port para o Core e acabei fazendo a comunicação com a SEFAZ na mão, atualmente eu uso apenas a impressão da DACTE e DANFE do projeto, então não tenho muito o que mostrar (além do que eu fiz um "cherry pick" dos projetos que me interessavam e isso acabou bagunçando os assemblies do FRX) 

A Microsoft já [deixou claro](https://devblogs.microsoft.com/dotnet/net-core-is-the-future-of-net/) que o futuro do .NET é o Core. Rodar o .NET Core como um microserviço "standalone" no Docker com Kestrel Web Server e programação assíncrona é algo bonito de ver para quem aprecia uma boa arquitetura e performance. Não conheço a história e objetivos desse projeto, mas se houver oportunidade de focar nesse tipo de arquitetura eu recomendo fortemente, talvez até mesmo um fork para não ficar muito preso à compatibilidade

marcosgerene commented on NFe.Wsdl.Standard/Autorizacao/AN/NfeRetAutorizacao4AN.cs:
```diff
@@ -0,0 +1,66 @@
+        }
+
+        /// <summary>
+        /// Classe base para a serealização no formato do envelope SOAP.
```
serialização*

danilobreda replied:
Alterados 27 erros de português incluindo esse.

marcosgerene commented:
@danilobreda 

Finalmente tive um tempo para ver, mesmo que superficialmente o código.

Em primeiro lugar meus parabéns pelo grande esforço que vem fazendo, está ficando show de bola!

Tenho algumas dúvidas, me perdoe se parecer algo óbvio, mas quero me manter por dentro do que está acontecendo...

Lá vai:

1. Você mantem por compatibilidade o atributo ```nfeCabecMsg```, mas na implementação toda coloca tudo que não seja 4.0 como não implementado... qual o motivo para ainda manter esse atributo?

2. Alguns métodos não foram implementados, exemplo distribuição, download... você pretende implementar esses métodos "faltantes"? digo somente os que ainda são validos, descontinuados como autorizacao3 não são necessários.

danilobreda commented:
@marcosgerene @aldoir 
Atualmente o projeto no qual estou utilizando esse pull request é um emissor de NFe com Cancelamento, Inutilização, Carta de Correção, Impressão de Danfe, Danfe de evento de cancelamento, carta de correção. Além da consulta de Status do NFE o subir o Certificado, para valida-lo.
Tudo isso desenvolvido em .net core 2.2 em ambiente WINDOWS. E rodando em produção em LINUX (Utilizando docker e imagens do aspnetcore da microsoft). Atualmente rodo esses containers em paralelo em uma arquitetura voltada a microservices. 
Todo o deploy (build e execução de testes) são feitos de forma automatizada em ambiente linux.
Já estamos com 2 clientes utilizando sem problemas até agora.
Além de vários testes efetuados sem nenhum problema.

O único problema que tive é que tive que gerar uns nugets por mim mesmo... O que acho que deveriamos gerar o nuget oficial, subir e eu utilizar nesse projeto e verificar se funciona tudo tranquilamente.

Outro ponto é que tive alguns problemas ao longo do caminho com a plataforma linux e docker, sendo resolvido facilmente com as seguintes linhas no dockerfile:
```
# Instala pacotes necessarios para danfe (libgdiplux)
RUN apt-get update \
    && apt-get install -y --no-install-recommends libgdiplus libc6-dev \
    && apt-get clean \
    && rm -rf /var/lib/apt/lists/*

# Instala fontes necessarios para danfe
RUN mkdir -p /usr/share/fonts/truetype/times
COPY fonts/* /usr/share/fonts/truetype/times/
# fontes instaladas: times.ttf, timesbd.ttf, timesbi.ttf, timesi.ttf
```
Outro problema foi o arquivos .frx e schemas sendo necessários na saída do build, pois a aplicação utiliza esses arquivos em runtime.

Acredito que esses problemas devem ser testados junto ao nuget final. **Por isso recomendo aguardar que o nuget seja gerado para a aprovação do pullrequest.** (Voces conseguiriam gerar o nuget @robertorp  @marcosgerene ?)

Também acho legal uma **documentação** para o pessoal saber o que fazer para usar no linux, pois como visto acima, existem algumas particularidades... Não sei como fazer isso, porem posso contribuir no texto.

danilobreda commented:
>     1. `nfeCabecMsg`

@marcosgerene realmente me deu dor de cabeça esse carinha kkk O problema é que existe uma interface antiga que é muito utilizada em vários lugares do sistema, principalmente nas versões antigas do NFe, a divisão de uma interface própria para a versão 4.0 e outra interface para antigas é recomendada porem não quis realizar nesse momento por conta de possíveis quebras e grandes alterações a nível de arquitetura do código. Eu tentei, mas achei muito complicado e desisti.

danilobreda commented:
>     2\. Alguns métodos não foram implementados, exemplo distribuição, download... você pretende implementar esses métodos "faltantes"? digo somente os que ainda são validos, descontinuados como autorizacao3 não são necessários.

@marcosgerene  não pretendo implementar no momento pois não conseguiria testar direito (nem sei realmente como funcionam e qual seu significado). Acredito que poderiamos ter uma documentação explicando quais são suportados e não são. Mas não é difícil implementar, eu só realmente não sei como testa-los.

marcosgerene commented:
@danilobreda 

>@marcosgerene não pretendo implementar no momento pois não conseguiria testar direito (nem sei realmente como funcionam e qual seu significado). Acredito que poderiamos ter uma documentação explicando quais são suportados e não são. Mas não é difícil implementar, eu só realmente não sei como testa-los

A minha aplicação implementa o DistribuiçãoDFe e o DownloadNFe (foram os que eu senti falta), se quiser posso te mandar a rotina em que uso os dois de forma combinada e a gente pensa em algo para o appteste que você fez...

----

**Update**

https://github.com/danilobreda/DFe.NET/pull/1

marcosgerene committed:
Merge branch 'nfecorereport' into nfecoreport

marcosgerene committed:
Merge pull request #38 from danilobreda/nfecoreport

Puxando para testar o core

marcosgerene committed:
NFeCore - AppTeste - Adicionado DistribuicaoDFe

https://github.com/ZeusAutomacao/DFe.NET/pull/1013#issuecomment-525918827

danilobreda committed:
faltando virgula em um enum

danilobreda committed:
Merge pull request #1 from marcosgerene/nfecorereport

NFeCore - AppTeste - Adicionado DistribuicaoDFe

danilobreda committed:
Merge remote-tracking branch 'origin/nfecoreport' into nfecoreport

danilobreda committed:
adicionado novos dados do generatereport para resolução de comflito

danilobreda committed:
Merge branch 'master' into nfecoreport

danilobreda committed:
Alteracao do readme e erros de português

danilobreda commented:
@marcosgerene 

- Alterado o Readme, com algumas dicas para impressão do Danfe em Linux e indicação do suporte ao .net core/standard nos projetos que suportam.
- Corrigido erros de português
- Resolução de conflitos 
- Merge da sua alteração no Appteste.

Estamos ready to go! 🥇

marcosgerene commented:
@danilobreda 
Poderia listar quais serviços não descontinuados que estão presentes no .Net Framework não estão presentes nessa branch? Ex: distribuição, downloadnfe, etc

danilobreda commented:
@marcosgerene começaram a ter muitos conflitos por alterações de namespace feito pelo @robertorp. Não sei como continuar com o pullrequest vivo daqui em diante. Está demorando muito para ser aprovado e cada vez mais começa a ter mais conflitos e trabalho para mim.
Temos que ver o que falta, como disse inicialmente é um pullrequest inicial para o dotnet core, com certeza não tem tudo que deveria ter, mas com o tempo devemos melhorar com outros pullrequests.

Pela grandeza das alterações feitas por esse pullrequest peço que @marcosgerene , @adeniltonbs  e @robertorp  priorezem a analise desse pullrequest.

danilobreda commented:
@marcosgerene  
**Todos do 4.0 estão suportados, alguns do 3.1 e alguns do 2.0 tambem.**
### Não suportados nesse branch: 
NFE Autorizacao 1.0
NFE Autorizacao 3.1
NFE Ret Autorizacao 1.0
NFE Ret Autorizacao 3.1
NFE Consulta Protocolo 1.0
NFE Consulta Protocolo 2.0
NFE Consulta Protocolo 3.1
NFeDistDFeInteresse 1.0
NFe Download NFE 1.0
(evento) RecepçãoEPEC 1.0
(evento) RecepçãoEvento 1.0
NFe Inutilizacao 1.0
NFe Inutilizacao 2.0
NFe Inutilizacao 3.1
NFe Recepcao 2.0
NFeRetRecepcao 2.0
NFe Status Servico 1.0
NFe Status Servico 3.1

marcosgerene commented:
@danilobreda 

Dos serviços listados, quais ainda estão "em vigor"?

Ex: NFE Autorizacao 1.0 e NFE Autorizacaoo 3.0 "não existem mais", certo?

Vamos tentar implementar o restante pra concluir o PR e pedir para o @robertorp ou o @adeniltonbs 
dar aquela  ultima olhada

danilobreda commented:
@marcosgerene [Neste link](http://www.nfe.fazenda.gov.br/portal/webServices.aspx?tipoConteudo=Wak0FwB7dKs=) temos todos que estão "em vigor" porem existem alguns na lista de versões antigas onde não estão mais funcionando, por conta da transição para 4.0.
Todas as 4.0 estão implementadas, e quais não existem a 4.0 existe versão anterior implementada.
Utilizo todas as APIs na aplicação da empresa e até agora esta tranquilo.

danilobreda committed:
alteracoes do nferetrato (arrumado conflitos)

danilobreda committed:
atualizado o relatorio retrato

danilobreda committed:
merge feito com master e resolvido conflitos nfecoreport

danilobreda commented:
Resolução de conflitos feita, merge feito do branch **danilobreda:nfecoreport** para **master**.
No wsdl foi alterado o namespace 
`using DFe.DocumentosEletronicos.Common;`
para
`using CTe.CTeOSDocumento.Common;`
Por conta das alterações seguintes no master: [28d2c793ce4951e2131f2f6209762706d2773590](https://github.com/ZeusAutomacao/DFe.NET/commit/28d2c793ce4951e2131f2f6209762706d2773590)

danilobreda committed:
Merge branch 'master' into nfecoreport

# Conflicts:
#	NFe.Danfe.Fast/NFe/NFeRetrato.frx
#	NFe.Wsdl/Autorizacao/NFeAutorizacao4.cs
#	Shared.DFe.Wsdl/Common/RequestBuilderAndSender.cs
#	Shared.DFe.Wsdl/Common/SoapUrls.cs
#	Shared.DFe.Wsdl/Common/SoapUtils.cs
#	Shared.DFe.Wsdl/Common/cteCabecMsg.cs
#	Shared.NFe.Servicos/ServicoNfeFactory.cs

marcosgerene commented:
@danilobreda 

Em alguns posts atrás você disse que não tinha download, manifestação, etc... então você fez, é isso?

ótimo! falei ontem com o Roberto, quero ver se a gnt consegue colocar esse PR em prática o quanto antes!

abraços

danilobreda commented:
@marcosgerene Todos que estão na versão 4.0 estão feitos, e alguns da 1.0 e 2.0.
O qual você disse não sei se são os abaixos, e:

Esse abaixo, que tem apenas no estado do Ceara não foi implementado.
**NfeDownloadNF** | 2.00 / 3.10 | https://nfe.sefaz.ce.gov.br/nfe2/services/NfeDownloadNF?wsdl
(Esse acredito que esta "deprecated" [https://nstecnologia.com.br/blog/substituicao-dos-webservices/](https://nstecnologia.com.br/blog/substituicao-dos-webservices/)

Esse abaixo que está no AN(ambiente nacional) também não foi implementado e é amplamente usado... esse é uma evolução do Download. 
**NFeDistribuicaoDFe** | 1.00 | https://www1.nfe.fazenda.gov.br/NFeDistribuicaoDFe/NFeDistribuicaoDFe.asmx

O NFeDistribuicaoDFE vou tentar implementar ainda hoje!
Desculpa minha confusão.

robertorp commented:
Atualmente estou em um evento, mas segunda feira vou dar total prioridade a esse pull request. Para evitar dores de cabeça com conflitos.

robertorp commented:
![image](https://user-images.githubusercontent.com/10251705/64643836-a4f5f780-d3e7-11e9-98ea-888518cb80a7.png)
Preciso de uma breve explicação, o projeto teste esta como .net core 2.1 mas o mesmo pede uma pasta 2.2 
essa pasta não é criada automaticamente , isso está correto?

danilobreda committed:
Atualizado projeto de teste AppTeste.NetCore para .net core 2.2 assim como atualmente é do Danfe.

danilobreda commented:
> 
> 
> ![image](https://user-images.githubusercontent.com/10251705/64643836-a4f5f780-d3e7-11e9-98ea-888518cb80a7.png)
> Preciso de uma breve explicação, o projeto teste esta como .net core 2.1 mas o mesmo pede uma pasta 2.2
> essa pasta não é criada automaticamente , isso está correto?

Foi um erro da versão do .net core. Arrumado.
Sobre essa questão de copiar o config, é o mesmo processo que o CTe utilizad, alias maioria dos processos feitos nesse pullrequest foi baseado no CTe

robertorp commented on DFe.Utils.Standard/CertificadoDigitaoUtil.cs:
```diff
@@ -0,0 +1,114 @@
+        /// </summary>
+        /// <param name="caminho">Caminho do arquivo + pastas</param>
+        /// <returns></returns>
+        private static string ArrumaCaminho(string caminho)
```
esse arruma caminho , não poderia ser um Path.Combine , não?

danilobreda replied:
Esse caminho vem da digitação do usuário, por isso é necessário esse "arruma". Essa classe utils é utilizada apenas para o console AppTeste.

robertorp commented on NFe.Danfe.AppTeste.NetCore/Funcoes.cs:
```diff
@@ -0,0 +1,60 @@
+        /// </summary>
+        /// <param name="caminho">Caminho do arquivo + pastas</param>
+        /// <returns></returns>
+        private static string ArrumaCaminho(string caminho)
```
esse arruma caminho , não poderia ser um Path.Combine , não?

danilobreda replied:
Esse caminho vem da digitação do usuário, por isso é necessário esse "arruma". Essa classe utils é utilizada apenas para o console AppTeste.

robertorp commented on NFe.Danfe.Fast.Standard/DanfeBase.cs:
```diff
@@ -0,0 +1,162 @@
+using Shared.NFe.Danfe.Base;
+using System.IO;
+
+namespace NFe.Danfe.Fast.Standard
```
Foi criado uma classe para cada projeto de impressão, ou seja tem duas impressões hoje? Uma para o .net core é uma para o .net framework ?

danilobreda replied:
O framework utilizados para impressão são diferentes (nuget FastResport.OpenSource para o standard e a dll FastReport para o .net framework)
A implementação das 2 é bem parecidas, porem alguns detalhes como o não suporte a impressão direta na impressora por exemplo fez com que a separação fosse necessária por enquanto.
Um problema foi a necessidade de 2 relatorios separados (.frx) por conta das classes no cabeçalho do arquivo frx serem diferentes. @marcosgerene disse para deixar os 2 separados por enquanto, que não teria problemas. Foi como segui.

robertorp commented on NFe.Danfe.Fast.Standard/NFCe/DanfeFrNfce.cs:
```diff
@@ -0,0 +1,73 @@
+    /// <summary>
+    /// Classe responsável pela impressão do DANFE da NFCe em Fast Reports
+    /// </summary>
+    public class DanfeFrNfce : DanfeBase
```
Foi criado uma classe para cada projeto de impressão, ou seja tem duas impressões hoje? Uma para o .net core é uma para o .net framework ?

danilobreda replied:
Respondido acima

robertorp commented:
não vou marcar todas classes que estão duplicadas, mas na parte de impressão tem muitas, foi realmente necessário, por qual motivo?

robertorp commented:
@marcosgerene  @danilobreda 
Está atualizado com as ultimas alterações também que fiz? Digo fizeram o merge com sucesso nesse ponto de tudo...
![image](https://user-images.githubusercontent.com/10251705/64869936-c0941480-d618-11e9-8945-b8e967af185f.png)

robertorp commented:
Pelo que observei, ficou dois projetos NFe.WSDL um que já estava pronto que é o do .netFramework e outro projeto para o .net Standard né

Um independente do outro pelo que ví 
Bom eu tinha pensado na época em fazer os dois projetos em um único projeto, mas isso posso estar fazendo com o tempo também hehe

robertorp commented on NFe.Wsdl/Autorizacao/NFeAutorizacao4.cs:
```diff
@@ -1,19 +1,31 @@
-﻿using System;
-using System.Xml;
-using CTe.CTeOSDocumento.Wsdl;
+﻿using CTe.CTeOSDocumento.Wsdl;
```
algo a ser ajustado com o tempo esse namespace.. por agora ótimo.  mas é algo a ser ajustado deixando esse comentário somente para eu mesmo me lembrar dele.

robertorp commented on Shared.CTe.Classes/Informacoes/Tipos/IndicadorPagamento.cs:
```diff
@@ -1,43 +1,43 @@
     ///     Indicador da forma de pagamento
-    ///     <para>0 � pagamento � vista;</para>
-    ///     <para>1 � pagamento � prazo;</para>
+    ///     <para>0  pagamento à vista;</para>
```
esse problema de - para um quadrado é por causa do tipo de arquivo né? UTF8 ou ASCII essas coisas, correto?

verifiquei que teve muitos mas muitos locais com esse tipo de coisa hehe mas isso é o de menos

danilobreda replied:
Eu tive que converter praticamente todo o projeto de ASCII para UTF8 pois no linux acaba que tendo muitos problemas de conversão, vivi na pele com problemas da empresa... e fiz uma ferramenta que converte de forma fácil. Acredito que o traço deveria existir, vou tentar voltar ele.

danilobreda replied:
Arrumado

robertorp commented:
Depois do almoço continuarei olhando, faltam apenas 87 arquivos.

danilobreda commented:
@robertorp 
> 
> 
> @marcosgerene @danilobreda
> Está atualizado com as ultimas alterações também que fiz? Digo fizeram o merge com sucesso nesse ponto de tudo...
> ![image](https://user-images.githubusercontent.com/10251705/64869936-c0941480-d618-11e9-8945-b8e967af185f.png)

Sim foi feito o merge por mim. Esta 100%



> 
> 
> não vou marcar todas classes que estão duplicadas, mas na parte de impressão tem muitas, foi realmente necessário, por qual motivo?

A duplicação de código não deve ocorrer, como respondi acima, foi necessário a separação pois são frameworks diferentes, o que consegui deixar no mesmo código deixei! criando um projeto shared e jogando lá o que for igual, por exemplo a classe **DanfeSharedHelper** O DanfeBase são diferentes por exemplo. Não tive como fazer igual! É como se fosse um novo projeto, feito para .netstandard. O Danfe não tinha como converter. Porem utilizam códigos base compartilhados.

danilobreda committed:
Voltado traço de summary do CTe.Classes.Informacoes.Tipos

robertorp commented:
Bom uma nova etapa concluída, @danilobreda,  Obrigado pela ajuda gigante!

robertorp commented:
@danilobreda , feito o merge! obrigado! 
vou estar ajustando o nuget para ter as duas versões !
